File: Salesman.hs
import Control.Arrow
import Data.Array
import Data.Maybe (mapMaybe)
import Text.Read (readMaybe)
import UCT

readCityLine :: [String] -> Maybe (Int, Int, Int)
readCityLine [i, x, y] = do
  i' <- readMaybe i
  x' <- readMaybe x
  y' <- readMaybe y
  return (i', x', y')
readCityLine _ = Nothing

readCityFile = mapMaybe (readCityLine . words) . lines

distances cities = array ((1, 1), (48, 48)) $
  concatMap (\ (i1, x1, y1) ->
    map (\ (i2, x2, y2) ->
      ((i1, i2), sqrt ((fI x1 - fI x2)^2 + (fI y1 - fI y2)^2))
    ) cities) cities
  where fI = fromIntegral

stateActs dists n visited
  | lv  < n = [(c : visited, score c) | c <- [1..n], c `notElem` visited]
  | lv == n = [(1 : visited, score 1)]
  | otherwise = []
  where
    score c = dists (head visited) c
    lv = length visited

tripDistance dists visited =
  sum $ map (dists !) $ zip visited (tail visited)

-- According to: http://elib.zib.de/pub/mp-testdata/tsp/tsplib/tsp/att48.opt.tour
bestSolution =
  [1,8,38,31,44,18,7,28,6,37,19,27,17,43,30,36,46,33,20,47,21,32,39,48,5,42,24,10,45,35,4,26,2,29,34,41,16,22,3,23,14,25,13,11,12,15,40,9,1]

main = do
  cities <- readCityFile <$> readFile "att48.tsp"
  let dists = distances cities
  let maxDist = maximum $ elems dists
  let distf c1 c2 = 1 - (dists ! (c1, c2)) / maxDist
  --print $ eval dists bestSolution
  let tspProblem = stateActs distf (length cities)
  let trees = take 15000 $ iterate (search tspProblem) $ emptyTree tspProblem [1]
  mapM_ (print . (tripDistance dists {- &&& id-}) . bestState byReward) trees

File: salesman.ml
open Batteries
open Uct

let read_city_line line =
  try Scanf.sscanf line " %d %d %d" (fun i x y -> Some (i, x, y))
  with Scanf.Scan_failure _ -> None

let read_cities fp = File.lines_of fp |> Enum.filter_map read_city_line |> List.of_enum

let distances cities =
  let n = List.length cities in
  let mat = Array.make_matrix n n 0.0 in
  let fi = float_of_int in
  List.iter (fun (c1, x1, y1) ->
    List.iter (fun (c2, x2, y2) ->
      mat.(c1-1).(c2-1) <- sqrt (((fi (x1 - x2)) ** 2. +. (fi (y1 - y2)) ** 2.))) cities) cities;
  mat

let rec butlast l = match l with
    [] -> []
  | x :: [] -> []
  | (x :: xs) -> x :: butlast xs

let round_trip_dist dists path =
  List.fold_right2 (fun c1 c2 -> (+.) (dists c1 c2)) (butlast path) (List.tl path) 0.

let state_acts dists n visited =
  let score c = dists (List.hd visited) c in
  let l = List.length visited in
  if l < n then
    (1--n) |> Enum.filter (fun c -> not (List.mem c visited))
    |> Enum.map (fun c -> (c :: visited, score c)) |> List.of_enum
  else if l = n then [(1 :: visited, score 1)]
  else []


let dupl x = (x, x)

let rec ltake n l = lazy (
  if n = 0 then LazyList.Nil
  else match LazyList.get l with
      Some (x, rest) -> LazyList.Cons (x, ltake (n-1) rest)
    | None -> LazyList.Nil)

let _ =
  let cities = read_cities "att48.tsp" in
  let dists = distances cities in
  let max_dist = Array.fold_left (Array.fold_left max) 0. dists in
  let distsf c1 c2 = dists.(c1-1).(c2-1) in
  let reward c1 c2 = 1. -. (distsf c1 c2) /. max_dist in
  let tsp_problem = state_acts reward (List.length cities) in
  let init = empty_tree tsp_problem [1] in
  let trees = LazyList.from_loop init (search tsp_problem %> dupl) in
  trees
  |> ltake 15000
  |> LazyList.iter (best_state by_reward %> round_trip_dist distsf %> Format.printf "%f\n")
